### ISSUE TITLE
Compiled code using pthreads breaks Vite bundler

### ISSUE THREAD

**Issue by jamsinclair** [emscripten-core/emscripten#22394]
**Description**

When code using threads is compiled with Emscripten and then bundled with Vite it does not bundle and throws the following Vite error:

```
[vite:worker-import-meta-url] Vite is unable to parse the worker options as the value is not static.To ignore this error, please use /* @vite-ignore */ in the worker options.
file
```

Vite has required static worker options for some time, over 2 years. Recent changes to Emscripten have made output code incompatible with the Vite bundler.

**Version of emscripten/emsdk:**

Affects versions >= `v3.1.58`

**Reproduction:**
I have created a repository at https://github.com/jamsinclair/emscripten-worker-options-vite-reproduction which show cases the problem.

**Potential Remedy:**

Refactor [library_pthread.js](https://github.com/emscripten-core/emscripten/blob/main/src/library_pthread.js) to use static values for the worker options. It'll be more verbose, but, it'll make the code compatible with Vite and maintain compatibility with other bundlers.

**Comment by jamsinclair** [comment]
Vite users could potentially write a simple rollup or vite plugin to work around this. Although, if we could fix this in Emscripten itself, that would be amazing 🙏.

I know emscripten is in a tricky spot trying to support bundlers and their different nuances 😅

**Comment by x0YYYY0** [comment]
I have run into this issue as well. Thank you for posting about it.  This is also unfortunate for those of us who are looking to use just shared memory, as the pthreads build option is the most comprehensive build option for that use case.

**Comment by sbc100** [comment]
Yup, we can/should fix this.  I wonder if we can write a test for it so we don't regress again.

Kind of still that vite can't see that the options are actually static.. just declared in a non-escaping var right before I think.

**Comment by dr-matt** [comment]
I've also run into this issue recently. Is there currently a workaround?

**Comment by zamfofex-dummy** [comment]
You can just use version 3.1.57, and it should work.

**Comment by sbc100** [comment]
We should fix this.  I'll try to add some tests to emscripten so that we don't regress again too.

**Comment by dr-matt** [comment]
@zamfofex-dummy thank you, good to know that's the version just before the regression. Unfortunately it would be tricky to move back to that version, as we're relying on other more recent changes.

@sbc100 that would be fantastic. Interestingly, we have one use case that works fine, but adding `-g2` to the build produces this error. Haven't dug into why yet, might be project-specific.

**Comment by jamsinclair** [comment]
@dr-matt I've scaffolded a quick attempt at a plugin that could patch this issue. Check this gist  https://gist.github.com/jamsinclair/6ad148d0590291077a4ce389c2b274ea.

Let me know how you go with it 👍

**Comment by kleisauke** [comment]
> Interestingly, we have one use case that works fine, but adding `-g2` to the build produces this error. Haven't dug into why yet, might be project-specific.

I think this is due to the Closure compiler (`--closure 1`) is automagically inlining the `workerOptions` variable. For example, I see this in the generated JS output:
```javascript
function Pe(){var a=new Worker(new URL("vips-es6.js",import.meta.url),{type:"module",name:"em-pthread"});Le.push(a)}
```

`-g2` always disables the use of the Closure compiler.
https://github.com/emscripten-core/emscripten/blob/e6ed9d3fc2821a72cc8ded4a02cac72b9f306f88/tools/link.py#L952-L954

**Comment by jamsinclair** [comment]
@mkl-ableton yes, upgrading to Vite >= v6.1.0 should hopefully solve your issue 🤞🏻

Context: Vite relaxed allowing dynamic values in WorkerOptions so long as it's not the `type` field.

**Comment by sbc100** [comment]
Thats great news @jamsinclair .  @mkl-ableton if updating vite doesn't work for you please reply and we can re-open.
